### ISSUE TITLE
Bug/Feature Request: Missing newline (`lf`) character at the end, if output target is file

### ISSUE THREAD

**Issue by stockbal** [SAP/abap-cleaner#124]
Hi,

If working with abapGit repositories, each ABAP file is always ended with a newline (`lf`) character. The abapGit report adds this newline behind the scene, as the web editors like GitHub are doing the same thing.

ABAP Cleaner currently does not do this, and so I would have always at least this one diff if I clean a whole lot of files.

![](https://user-images.githubusercontent.com/35834861/271705222-b24ac263-244d-48e8-afda-559453cf7492.png)

The code adjustment would be very simple. I am not sure if an extra cli option is required for this, as a newline at the end of file will not hurt anyone - I think 😉.

Kind regards,
Ludwig

**Comment by jmgrassau** [comment]
Hi Ludwig, 

just to be sure I understand this correctly: This is just meant for the case of command line usage, i.e. for FrmMain.writeCleanUpResult(), right? Because I think in other contexts (e.g. returning the cleanup result to the ADT editor), the final LF would not be helpful.

And would "cr lf" (\\r\\n) be fine as well? Because this is always used between ABAP statements (ABAP.LINE_SEPARATOR). 

Kind regards,
Jörg-Michael

**Comment by stockbal** [comment]
Hi Jörg-Michael,

correct. This is only for cli usage.
Regarding `crlf` vs `lf`. I think it has to be `lf`. Then ABAP Clean would be conform with [abapGit] (https://github.com/abapGit/abapGit/blob/0ab304420bdeba11e5a8e188fa9cb282c7ba7546/src/objects/core/zcl_abapgit_objects_files.clas.abap#L145C1-L146) (cl_abap_char_utilities=>newline = `lf`).

I will check if `crlf` will cause any problems or not.

Kind regards,
Ludwig

**Comment by jmgrassau** [comment]
Hi Ludwig, 

okay, I enhanced this to ```persistency.writeAllTextToFile(targetPath, output + ABAP.LINE_SEPARATOR);```

Regarding ```crlf``` versus ```lf```, I think I chose ```crlf``` because that is what ABAP cleaner gets from the ADT editor (if I remember correctly), at least on Windows. 

Kind regards,
Jörg-Michael

**Comment by stockbal** [comment]
Hi Jörg-Michael,

I also checked if it would make a difference with `crlf` or not. I couldn't detect one. At least no visible diff from git. Could also be because of the git setting `core.autocrlf = true`. So I think it should be fine either way 🤞.

Kind regards,
Ludwig

**Comment by jmgrassau** [comment]
Hi Ludwig, 

great, thanks for checking! Actually, I realized that the method must check whether a ```lf``` or ```crlf``` is missing at the end, and only then add it; otherwise, if the source file ends with 3 line feeds, you'd get 4 line feeds in the target file etc. (this is because the parser ignores a single line feed at the end, but not multiple line feeds).

I also enhanced this method a bit to write the "sourceFolderFile" to the console if(!) ```--stats``` and/or ```--usedrules``` are requested. Otherwise you wouldn't know which file the respective information belongs to. 

Kind regards,
Jörg-Michael

Kind regards,
Jörg-Michael
